File: android/app/src/main/java/com/cactus/remoteterminal/protocol/Incoming.kt
package com.cactus.remoteterminal.protocol

import org.json.JSONObject

/** Relay → phone messages, parsed into typed events (PROTOCOL.md §5). */
sealed class RelayEvent {
    data class Welcome(
        val version: Int, val connId: String, val accountId: String, val deviceId: String,
        val caps: List<String>, val limits: Limits, val agents: List<AgentInfo>, val devices: List<DeviceInfo>,
    ) : RelayEvent()
    data class AgentList(val agents: List<AgentInfo>) : RelayEvent()
    data class AgentOnline(val agent: AgentInfo) : RelayEvent()
    data class AgentOffline(val agentId: String, val lastSeen: Long?) : RelayEvent()
    data class AgentUpdated(val agentId: String, val name: String) : RelayEvent()
    data class AgentRemoved(val agentId: String, val by: String?) : RelayEvent()
    data class DeviceList(val devices: List<DeviceInfo>) : RelayEvent()
    data class DeviceUpdated(val deviceId: String, val name: String) : RelayEvent()
    data class DeviceRevoked(val deviceId: String, val by: String?) : RelayEvent()
    data class SessionCreated(val agentId: String, val session: SessionInfo, val reqId: String?) : RelayEvent()
    data class SessionAttached(val agentId: String, val sessionId: String, val from: Long, val seq: Long, val cols: Int, val rows: Int, val reqId: String?) : RelayEvent()
    data class SessionUpdated(val agentId: String, val sessionId: String, val title: String?, val state: String?, val cols: Int?, val rows: Int?, val exitCode: Int?) : RelayEvent()
    data class Exit(val agentId: String, val sessionId: String, val code: Int?) : RelayEvent()
    data class SessionClosed(val agentId: String, val sessionId: String, val reason: String) : RelayEvent()
    data class SessionLag(val agentId: String, val sessionId: String) : RelayEvent()
    data class Output(val agentId: String, val sessionId: String, val seq: Long, val data: String) : RelayEvent()
    data class FileStored(val agentId: String, val sessionId: String, val path: String, val size: Long, val reqId: String?) : RelayEvent()
    data class Error(val code: String, val message: String, val reqId: String?, val agentId: String?, val sessionId: String?) : RelayEvent() {
        /** Human-readable text for the UI. */
        val display: String
            get() = when (code) {
                "agent_offline" -> "The machine is offline."
                "unknown_session" -> "That terminal no longer exists."
                "limit_reached" -> "Limit reached: $message"
                "rate_limited" -> "Too many requests; try again in a moment."
                "forbidden" -> "Not allowed: $message"
                "unauthorized" -> "Not authorized. Pair this phone again."
                "unsupported_version" -> "The relay speaks a different protocol version."
                else -> message.ifEmpty { code }
            }
    }
    object Pong : RelayEvent()
    data class Unknown(val type: String) : RelayEvent()
}

object Incoming {
    private fun JSONObject.str(key: String): String? = if (!has(key) || isNull(key)) null else optString(key)
    private fun JSONObject.int(key: String): Int? = if (!has(key) || isNull(key)) null else optInt(key)
    private fun JSONObject.long(key: String): Long? = if (!has(key) || isNull(key)) null else optLong(key)

    /** Parse one relay message. Throws on malformed JSON; unknown types become [RelayEvent.Unknown]. */
    fun parse(text: String): RelayEvent {
        val o = JSONObject(text)
        return when (val type = o.optString("type")) {
            "output" -> RelayEvent.Output(o.getString("agent"), o.getString("session"), o.getLong("seq"), o.optString("data"))
            "welcome" -> RelayEvent.Welcome(
                version = o.optInt("v", 0), connId = o.optString("connId"), accountId = o.optString("accountId"),
                deviceId = o.optString("deviceId"), caps = o.optJSONArray("caps").toStrings(),
                limits = Limits.fromJson(o.optJSONObject("limits")),
                agents = o.optJSONArray("agents").toList { AgentInfo.fromJson(it) },
                devices = o.optJSONArray("devices").toList { DeviceInfo.fromJson(it) },
            )
            "agent.list" -> RelayEvent.AgentList(o.optJSONArray("agents").toList { AgentInfo.fromJson(it) })
            "agent.online" -> RelayEvent.AgentOnline(AgentInfo.fromJson(o.getJSONObject("agent")))
            "agent.offline" -> RelayEvent.AgentOffline(o.getString("agent"), o.long("lastSeen"))
            "agent.updated" -> RelayEvent.AgentUpdated(o.getString("agent"), o.optString("name"))
            "agent.removed" -> RelayEvent.AgentRemoved(o.getString("agent"), o.str("by"))
            "device.list" -> RelayEvent.DeviceList(o.optJSONArray("devices").toList { DeviceInfo.fromJson(it) })
            "device.updated" -> RelayEvent.DeviceUpdated(o.getString("device"), o.optString("name"))
            "device.revoked" -> RelayEvent.DeviceRevoked(o.getString("device"), o.str("by"))
            "session.created" -> RelayEvent.SessionCreated(o.getString("agent"), SessionInfo.fromJson(o.getJSONObject("session")), o.str("reqId"))
            "session.attached" -> RelayEvent.SessionAttached(
                o.getString("agent"), o.getString("session"), o.getLong("from"), o.getLong("seq"), o.optInt("cols"), o.optInt("rows"), o.str("reqId"),
            )
            "session.updated" -> RelayEvent.SessionUpdated(
                o.getString("agent"), o.getString("session"), o.str("title"), o.str("state"), o.int("cols"), o.int("rows"), o.int("exitCode"),
            )
            "exit" -> RelayEvent.Exit(o.getString("agent"), o.getString("session"), o.int("code"))
            "session.closed" -> RelayEvent.SessionClosed(o.getString("agent"), o.getString("session"), o.optString("reason", "closed"))
            "session.lag" -> RelayEvent.SessionLag(o.getString("agent"), o.getString("session"))
            "file.stored" -> RelayEvent.FileStored(o.getString("agent"), o.getString("session"), o.optString("path"), o.optLong("size", 0L), o.str("reqId"))
            "error" -> RelayEvent.Error(o.optString("code", "internal"), o.optString("message"), o.str("reqId"), o.str("agent"), o.str("session"))
            "pong" -> RelayEvent.Pong
            else -> RelayEvent.Unknown(type)
        }
    }
}

File: desktop/ui/js/protocol/incoming.js
/**
 * Relay → client messages, parsed into typed events (PROTOCOL.md §5).
 * A port of `Incoming.kt`.
 *
 * Every event is `{ kind, ... }`; unknown types become `{ kind: 'unknown' }`
 * rather than an error, because the protocol says unknown fields and
 * capabilities are ignored.
 */

import { agentFromJson, deviceFromJson, sessionFromJson, metricsFromJson, limitsFromJson } from './messages.js';

const optStr = (o, key) => (o[key] == null ? null : String(o[key]));
const optNum = (o, key) => (typeof o[key] === 'number' && Number.isFinite(o[key]) ? o[key] : null);
const list = (v, map) => (Array.isArray(v) ? v.map(map) : []);

/** Human-readable text for the UI, mirroring `RelayEvent.Error.display`. */
export function errorDisplay(event) {
  switch (event.code) {
    case 'agent_offline': return 'The machine is offline.';
    case 'unknown_session': return 'That terminal no longer exists.';
    case 'limit_reached': return `Limit reached: ${event.message}`;
    case 'rate_limited': return 'Too many requests; try again in a moment.';
    case 'forbidden': return `Not allowed: ${event.message}`;
    case 'unauthorized': return 'Not authorized. Pair this device again.';
    case 'unsupported_version': return 'The relay speaks a different protocol version.';
    default: return event.message || event.code;
  }
}

export function makeError(code, message, reqId = null, agentId = null, sessionId = null) {
  return { kind: 'error', code, message, reqId, agentId, sessionId };
}

/** Parse one relay message. Throws on malformed JSON. */
export function parseIncoming(text) {
  const o = JSON.parse(text);
  if (!o || typeof o !== 'object') throw new Error('not an object');
  switch (o.type) {
    case 'output':
      return { kind: 'output', agentId: String(o.agent), sessionId: String(o.session), seq: Number(o.seq), data: o.data == null ? '' : String(o.data) };
    case 'welcome':
      return {
        kind: 'welcome',
        version: Number(o.v) || 0,
        connId: String(o.connId ?? ''),
        accountId: String(o.accountId ?? ''),
        deviceId: String(o.deviceId ?? ''),
        caps: list(o.caps, String),
        limits: limitsFromJson(o.limits),
        agents: list(o.agents, agentFromJson),
        devices: list(o.devices, deviceFromJson),
      };
    case 'agent.list':
      return { kind: 'agentList', agents: list(o.agents, agentFromJson) };
    case 'agent.online':
      return { kind: 'agentOnline', agent: agentFromJson(o.agent) };
    case 'agent.offline':
      return { kind: 'agentOffline', agentId: String(o.agent), lastSeen: optNum(o, 'lastSeen') };
    case 'agent.updated':
      return { kind: 'agentUpdated', agentId: String(o.agent), name: String(o.name ?? '') };
    case 'agent.metrics':
      return { kind: 'agentMetrics', agentId: String(o.agent), metrics: metricsFromJson(o.metrics) };
    case 'agent.removed':
      return { kind: 'agentRemoved', agentId: String(o.agent), by: optStr(o, 'by') };
    case 'device.list':
      return { kind: 'deviceList', devices: list(o.devices, deviceFromJson) };
    case 'device.updated':
      return { kind: 'deviceUpdated', deviceId: String(o.device), name: String(o.name ?? '') };
    case 'device.revoked':
      return { kind: 'deviceRevoked', deviceId: String(o.device), by: optStr(o, 'by') };
    case 'session.created':
      return { kind: 'sessionCreated', agentId: String(o.agent), session: sessionFromJson(o.session), reqId: optStr(o, 'reqId') };
    case 'session.attached':
      return {
        kind: 'sessionAttached',
        agentId: String(o.agent),
        sessionId: String(o.session),
        from: Number(o.from) || 0,
        seq: Number(o.seq) || 0,
        cols: Number(o.cols) || 0,
        rows: Number(o.rows) || 0,
        reqId: optStr(o, 'reqId'),
      };
    case 'session.updated':
      return {
        kind: 'sessionUpdated',
        agentId: String(o.agent),
        sessionId: String(o.session),
        title: optStr(o, 'title'),
        state: optStr(o, 'state'),
        cols: optNum(o, 'cols'),
        rows: optNum(o, 'rows'),
        exitCode: optNum(o, 'exitCode'),
        cwd: optStr(o, 'cwd'),
      };
    case 'exit':
      return { kind: 'exit', agentId: String(o.agent), sessionId: String(o.session), code: optNum(o, 'code') };
    case 'session.closed':
      return { kind: 'sessionClosed', agentId: String(o.agent), sessionId: String(o.session), reason: String(o.reason ?? 'closed') };
    case 'session.lag':
      return { kind: 'sessionLag', agentId: String(o.agent), sessionId: String(o.session) };
    case 'file.stored':
      return {
        kind: 'fileStored',
        agentId: String(o.agent),
        sessionId: String(o.session),
        path: String(o.path ?? ''),
        size: Number(o.size) || 0,
        reqId: optStr(o, 'reqId'),
      };
    case 'error':
      return makeError(String(o.code ?? 'internal'), String(o.message ?? ''), optStr(o, 'reqId'), optStr(o, 'agent'), optStr(o, 'session'));
    case 'pong':
      return { kind: 'pong' };
    default:
      return { kind: 'unknown', type: String(o.type ?? '') };
  }
}
